src/libpmemobj/ctl.c

```diff
@@ -0,0 +1,668 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * ctl.c -- implementation of the interface for examining and modification of
+ *	the library internal state
+ */
+
+#include <sys/param.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include "libpmem.h"
+#include "libpmemobj.h"
+
+#include "util.h"
+#include "out.h"
+#include "lane.h"
+#include "redo.h"
+#include "memops.h"
+#include "pmalloc.h"
+#include "heap_layout.h"
+#include "list.h"
+#include "cuckoo.h"
+#include "ctree.h"
+#include "obj.h"
+#include "sync.h"
+#include "valgrind_internal.h"
+#include "ctl.h"
+#include "memblock.h"
+#include "heap.h"
+
+#define CTL_MAX_ENTRIES 100
+
+#define MAX_CONFIG_FILE_LEN (1 << 20) /* 1 megabyte */
+
+#define CTL_STRING_QUERY_SEPARATOR ";"
+#define CTL_NAME_VALUE_SEPARATOR "="
+#define CTL_QUERY_NODE_SEPARATOR "."
+#define CTL_VALUE_ARG_SEPARATOR ","
+
+static int ctl_global_first_free = 0;
+static struct ctl_node CTL_NODE(global)[CTL_MAX_ENTRIES];
+
+/*
+ * This is the top level node of the ctl tree structure. Each node can contain
+ * children and leaf nodes.
+ *
+ * Internal nodes simply create a new path in the tree whereas child nodes are
+ * the ones providing the read/write functionality by the means of callbacks.
+ *
+ * Each tree node must be NULL-terminated, CTL_NODE_END macro is provided for
+ * convience.
+ */
+struct ctl {
+	struct ctl_node root[CTL_MAX_ENTRIES];
+	int first_free;
+};
+
+/*
+ * String provider is the simplest, elementary, query provider. It can be used
+ * directly to parse environment variables or in conjuction with other code to
+ * provide more complex behavior. It is initialized with a string containing all
+ * of the queries and tokenizes it into separate structures.
+ */
+struct ctl_string_provider {
+	struct ctl_query_provider super;
+
+	char *buf; /* stores the entire string that needs to be parsed */
+	char *sptr; /* for internal use of strtok */
+};
+
+/*
+ * File provider builts on top of the string provider to facilitate reading
+ * query data from a user-provided file.
+ */
+struct ctl_file_provider {
+	struct ctl_string_provider super;
+	FILE *config;
+};
+
+/*
+ * ctl_find_node -- (internal) searches for a matching entry point in the
+ *	provided nodes
+ *
+ * The caller is responsible for freeing all of the allocated indexes,
+ * regardless of the return value.
+ */
+static struct ctl_node *
+ctl_find_node(struct ctl_node *nodes, const char *name,
+	struct ctl_indexes *indexes)
+{
+	struct ctl_node *n = NULL;
+	char *sptr = NULL;
+	char *parse_str = Strdup(name);
+	if (parse_str == NULL)
+		return NULL;
+
+	char *node_name = strtok_r(parse_str, CTL_QUERY_NODE_SEPARATOR, &sptr);
+
+	/*
+	 * Go through the string and separate tokens that correspond to nodes
+	 * in the main ctl tree.
+	 */
+	while (node_name != NULL) {
```

Comment:
Not sure if I understand. The only way I could move the strtok somehow inside the loop logic is to change this while into a for. It would be more complicated.